pkg/cri/server/image_pull.go

```diff
@@ -786,3 +795,29 @@ func (rt *pullRequestReporterRoundTripper) RoundTrip(req *http.Request) (*http.R
 	}
 	return resp, err
 }
+
+// Given that runtime information is not passed from PullImageRequest, we depend on an experimental annotation
+// passed from pod sandbox config to get the runtimeHandler. The annotation key is specified in configuration.
+// Once we know the runtime, try to override default snapshotter if it is set for this runtime.
+// See https://github.com/containerd/containerd/issues/6657
+func (c *criService) snapshotterFromPodSandboxConfig(ctx context.Context, imageRef string,
+	s *runtime.PodSandboxConfig) (string, error) {
+	snapshotter := c.config.ContainerdConfig.Snapshotter
+	if s == nil || s.Annotations == nil {
+		return snapshotter, nil
+	}
+
+	runtimeHandler, ok := s.Annotations[annotations.RuntimeHandler]
+	if !ok {
+		return snapshotter, nil
+	}
+
+	ociRuntime, err := c.getSandboxRuntime(s, runtimeHandler)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get sandbox runtime for %s", runtimeHandler)
+	}
+
+	snapshotter = c.runtimeSnapshotter(context.Background(), ociRuntime)
+	log.G(ctx).Infof("PullImage %s for runtime %s, using snapshotter %s", imageRef, runtimeHandler, snapshotter)
```

Comment:
"experimental: PullImage ref: %q ...

Reply:
please confirm the snapshotter is not nil and if nil make it an experimental: error otherwise this success info regarding retrieving the snapshotted from the runtime options..

Reply:
Added `experimental` mark

Regarding checking the snapshotter not empty, the `runtimeSnapshotter` already guaranteed that it's not empty, I've also added the UT to protect the future changes. 

```
func (c *criService) runtimeSnapshotter(ctx context.Context, ociRuntime criconfig.Runtime) string {
	if ociRuntime.Snapshotter == "" {
		return c.config.ContainerdConfig.Snapshotter
	}

	log.G(ctx).Infof("Set snapshotter for runtime %s to %s", ociRuntime.Type, ociRuntime.Snapshotter)
	return ociRuntime.Snapshotter
}
```

Reply:
sort of.. I agree calling it experimental was a bad idea..  If we have a validator for c.config.ContainerdConfig.Snapshotter not being nil somewhere and/or being a valid loaded snapshotter already, then yes we don't need to check here. We should be covered.  However, if we don't have a config validator for c.config.ContainerdConfig.Snapshotter.. then I see the opportunity for someone going with no default at all and instead setting snapshotter config on each runtime.  Then in that case.. perhaps a check / error here is warranted, at least if it is nil.  Thoughts?

Reply:
right so.. with the introduction of this feature.. we should no longer have a must have requirement for c.config.ContainerdConfig.Snapshotter to be non-nil rather, either c.config.ContainerdConfig.Snapshotter must be non-nil or at least if it is nil the runc runtime must have a non-nil snapshotter or .. we won't pass critest/e2e node buckets, our own cri integration etc..  We can carry that item (fixing the validator to allow nil for the global setting if the runc runtime (possibly all runtimes) has snapshotter defined) as one of the todo items for this feature. If you like or add it to this PR :-)

Reply:
I'm tracking this item as to refine snapshotter validation in the TODO https://github.com/containerd/containerd/issues/6657#issuecomment-1129385189 it might need a bit discussion and agreement so I'll leave current behavior unchanged in this PR.